In [ ]:
!pip install pyspark

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 281.3 MB 45 kB/s 
     |████████████████████████████████| 199 kB 65.6 MB/s 
  Created wheel for pyspark: filename=pyspark-3.3.0-py2.py3-none-any.whl size=281764026 sha256=f1d3f2224c8a0a4756a5893806d82ce89018ec79ccd524332245fe5effabe0aa
  Stored in directory: /root/.cache/pip/wheels/7a/8e/1b/f73a52650d2e5f337708d9f6a1750d451a7349a867f928b885
Successfully built pyspark


In [ ]:
import time

## Importing Dataset

In [ ]:
from pyspark.sql import SparkSession

spark = SparkSession.builder.appName('tree').getOrCreate()
df = spark.read.csv('/content/drive/MyDrive/Datasets/Thesis/dr16.csv', inferSchema=True, header=True)
df.printSchema()

root
 |-- _c0: integer (nullable = true)
 |-- objid: double (nullable = true)
 |-- modelMag_u: double (nullable = true)
 |-- modelMag_g: double (nullable = true)
 |-- modelMag_r: double (nullable = true)
 |-- modelMag_i: double (nullable = true)
 |-- modelMag_z: double (nullable = true)
 |-- fiberMag_u: double (nullable = true)
 |-- fiberMag_g: double (nullable = true)
 |-- fiberMag_r: double (nullable = true)
 |-- fiberMag_i: double (nullable = true)
 |-- fiberMag_z: double (nullable = true)
 |-- petroR50_r: double (nullable = true)
 |-- petroR90_r: double (nullable = true)
 |-- petroR50_z: double (nullable = true)
 |-- petroR90_z: double (nullable = true)
 |-- r: double (nullable = true)
 |-- i: double (nullable = true)
 |-- z: double (nullable = true)
 |-- redshift: double (nullable = true)
 |-- zerr: double (nullable = true)
 |-- mmug: double (nullable = true)
 |-- mmgr: double (nullable = true)
 |-- mmri: double (nullable = true)
 |-- mmiz: double (nullable = true)
 |-- mfug: doub

##Data Pre-Processing

In [ ]:
df = df.drop('objid', '_c0')
df.show(5)

+----------+----------+----------+----------+----------+----------+----------+----------+----------+----------+------------------+----------+------------------+----------+--------+--------+--------+---------+-----------------+---------+------------------+---------+--------+------------------+---------+--------+------------------+
|modelMag_u|modelMag_g|modelMag_r|modelMag_i|modelMag_z|fiberMag_u|fiberMag_g|fiberMag_r|fiberMag_i|fiberMag_z|        petroR50_r|petroR90_r|        petroR50_z|petroR90_z|       r|       i|       z| redshift|             zerr|     mmug|              mmgr|     mmri|    mmiz|              mfug|     mfgr|    mfri|              mfiz|
+----------+----------+----------+----------+----------+----------+----------+----------+----------+----------+------------------+----------+------------------+----------+--------+--------+--------+---------+-----------------+---------+------------------+---------+--------+------------------+---------+--------+------------------+
|  2

###Check Number of Rows and columns

In [ ]:
print(df.count(), len(df.columns))

3524477 27


### Remove Duplicate rows in Dataframe

In [ ]:
### Remove Duplicate rows in Dataframe
df = df.dropDuplicates()
print(df.count(), len(df.columns))

3524477 27


Since the number of rows remains the same, this means that there was no duplicate rows in Dataframe

In [ ]:
import pyspark.sql.functions as func

df = df.withColumn("redshift", func.round(df["redshift"], 2).cast('integer'))

In [ ]:
import pandas as pd
pd.DataFrame(df.take(5), columns = df.columns).transpose()

0          1          2          3          4
modelMag_u  21.322890  21.540440  24.099690  19.703640  21.891370
modelMag_g  21.199790  21.344430  21.893860  17.857980  19.901890
modelMag_r  20.978130  20.455390  20.295770  16.785280  18.379160
modelMag_i  20.786010  19.635970  19.441590  16.337970  17.835950
modelMag_z  20.602870  19.201110  19.201700  16.005240  17.496330
fiberMag_u  21.721530  22.420790  24.791400  20.956700  22.894670
fiberMag_g  21.571780  22.167020  22.645580  19.119800  20.870210
fiberMag_r  21.418020  21.318560  21.237360  18.061850  19.368600
fiberMag_i  21.290620  20.557740  20.370210  17.609200  18.812600
fiberMag_z  21.389900  20.050190  20.087840  17.276210  18.468090
petroR50_r   0.706643   1.324028   1.441734   1.968045   1.514206
petroR90_r   1.368751   2.498909   3.625134   6.430813   4.067373
petroR50_z   0.141109   1.282831   2.059727   1.948154   1.487422
petroR90_z   0.190463   2.290378   4.790148   6.577235   4.122421
r           20.978130  20.455390  20.295770  16.785280  18.379160
i           20.786010  19.635970  19.441590  16.337970  17.835950
z           20.602870  19.201110  19.201700  16.005240  17.496330
redshift     2.000000   0.000000   0.000000   0.000000   0.000000
zerr         0.000703   0.000098   0.000169   0.000026   0.000041
mmug        21.261340  21.442435  22.996775  18.780810  20.896630
mmgr        21.088960  20.899910  21.094815  17.321630  19.140525
mmri        20.882070  20.045680  19.868680  16.561625  18.107555
mmiz        20.694440  19.418540  19.321645  16.171605  17.666140
mfug        21.646655  22.293905  23.718490  20.038250  21.882440
mfgr        21.494900  21.742790  21.941470  18.590825  20.119405
mfri        21.354320  20.938150  20.803785  17.835525  19.090600
mfiz        21.340260  20.303965  20.229025  17.442705  18.640345

In [ ]:
df.show(5)

+----------+------------------+----------+------------------+------------------+----------+----------+----------+----------+----------+----------+----------+----------+-----------------+--------+------------------+------------------+--------+--------------------+---------+---------+---------+---------+------------------+---------+---------+---------+
|modelMag_u|        modelMag_g|modelMag_r|        modelMag_i|        modelMag_z|fiberMag_u|fiberMag_g|fiberMag_r|fiberMag_i|fiberMag_z|petroR50_r|petroR90_r|petroR50_z|       petroR90_z|       r|                 i|                 z|redshift|                zerr|     mmug|     mmgr|     mmri|     mmiz|              mfug|     mfgr|     mfri|     mfiz|
+----------+------------------+----------+------------------+------------------+----------+----------+----------+----------+----------+----------+----------+----------+-----------------+--------+------------------+------------------+--------+--------------------+---------+---------+---------+-

In [ ]:
df.columns

['modelMag_u',
 'modelMag_g',
 'modelMag_r',
 'modelMag_i',
 'modelMag_z',
 'fiberMag_u',
 'fiberMag_g',
 'fiberMag_r',
 'fiberMag_i',
 'fiberMag_z',
 'petroR50_r',
 'petroR90_r',
 'petroR50_z',
 'petroR90_z',
 'r',
 'i',
 'z',
 'redshift',
 'zerr',
 'mmug',
 'mmgr',
 'mmri',
 'mmiz',
 'mfug',
 'mfgr',
 'mfri',
 'mfiz']

In [ ]:
df.show(5)

+----------+------------------+----------+------------------+------------------+----------+----------+----------+----------+----------+----------+----------+----------+-----------------+--------+------------------+------------------+--------+--------------------+---------+---------+---------+---------+------------------+---------+---------+---------+
|modelMag_u|        modelMag_g|modelMag_r|        modelMag_i|        modelMag_z|fiberMag_u|fiberMag_g|fiberMag_r|fiberMag_i|fiberMag_z|petroR50_r|petroR90_r|petroR50_z|       petroR90_z|       r|                 i|                 z|redshift|                zerr|     mmug|     mmgr|     mmri|     mmiz|              mfug|     mfgr|     mfri|     mfiz|
+----------+------------------+----------+------------------+------------------+----------+----------+----------+----------+----------+----------+----------+----------+-----------------+--------+------------------+------------------+--------+--------------------+---------+---------+---------+-

In [ ]:
from pyspark.sql.functions import monotonically_increasing_id
df = df.withColumn("objid", monotonically_increasing_id())
df.show(5)

+----------+------------------+----------+------------------+------------------+----------+----------+----------+----------+----------+----------+----------+----------+-----------------+--------+------------------+------------------+--------+--------------------+---------+---------+---------+---------+------------------+---------+---------+---------+-----+
|modelMag_u|        modelMag_g|modelMag_r|        modelMag_i|        modelMag_z|fiberMag_u|fiberMag_g|fiberMag_r|fiberMag_i|fiberMag_z|petroR50_r|petroR90_r|petroR50_z|       petroR90_z|       r|                 i|                 z|redshift|                zerr|     mmug|     mmgr|     mmri|     mmiz|              mfug|     mfgr|     mfri|     mfiz|objid|
+----------+------------------+----------+------------------+------------------+----------+----------+----------+----------+----------+----------+----------+----------+-----------------+--------+------------------+------------------+--------+--------------------+---------+---------

In [ ]:
from pyspark.ml.feature import VectorAssembler

assembler = VectorAssembler(inputCols = ['modelMag_u', 'modelMag_g', 'modelMag_r', 'modelMag_i', 'modelMag_z', 'fiberMag_u', 'fiberMag_g', 'fiberMag_r', 'fiberMag_i', 'fiberMag_z', 'petroR50_r', 'petroR90_r', 'petroR50_z', 'petroR90_z', 'r', 'i', 'z', 'mmug', 'mmgr', 'mmri', 'mmiz', 'mfug', 'mfgr', 'mfri', 'mfiz'], outputCol = 'features')

In [ ]:
output = assembler.transform(df)

In [ ]:
final_df = output.select('features', 'redshift')
final_df.show(3)

+--------------------+--------+
|            features|redshift|
+--------------------+--------+
|[21.32289,21.1997...|       2|
|[21.54044,21.3444...|       0|
|[24.09969,21.8938...|       0|
+--------------------+--------+
only showing top 3 rows



In [ ]:
train, test = final_df.randomSplit([0.7, 0.3])

In [ ]:
from pyspark.ml.classification import (DecisionTreeClassifier, RandomForestClassifier, 
                                      GBTClassifier)
from pyspark.ml import Pipeline

dt = DecisionTreeClassifier(labelCol = 'redshift', featuresCol = 'features')
#rf = RandomForestClassifier(labelCol = 'redshift', featuresCol = 'features')

In [ ]:
start_time = time.time()
dt_model = dt.fit(train)
print("%s seconds" % (time.time() - start_time))

226.8268220424652 seconds


In [ ]:
#print("%s seconds" % (time.time() - start_time))

In [ ]:
dt_train_prediction = dt_model.transform(train)
dt_train_prediction.show(5)

+--------------------+--------+--------------------+--------------------+----------+
|            features|redshift|       rawPrediction|         probability|prediction|
+--------------------+--------+--------------------+--------------------+----------+
|[13.78523,12.4974...|       0|[1661635.0,16122....|[0.98662245088812...|       0.0|
|[13.8472700000000...|       0|[1661635.0,16122....|[0.98662245088812...|       0.0|
|[14.02077,12.3308...|       0|[1661635.0,16122....|[0.98662245088812...|       0.0|
|[14.1087700000000...|       0|[1661635.0,16122....|[0.98662245088812...|       0.0|
|[14.13117,12.5237...|       0|[1661635.0,16122....|[0.98662245088812...|       0.0|
+--------------------+--------+--------------------+--------------------+----------+
only showing top 5 rows



In [ ]:
from pyspark.ml.evaluation import MulticlassClassificationEvaluator
multi_evaluator = MulticlassClassificationEvaluator(labelCol = 'redshift', metricName = 'accuracy')
print('Random Forest Training Accuracy:', multi_evaluator.evaluate(dt_train_prediction))
print('Training data (weightedPrecision): ', multi_evaluator.setMetricName('weightedPrecision').evaluate(dt_train_prediction))

Random Forest Training Accuracy: 0.8615107549317915
Training data (weightedPrecision):  0.8668010624481051


In [ ]:
dt_predictions = dt_model.transform(test)

In [ ]:
from pyspark.ml.evaluation import MulticlassClassificationEvaluator

multi_evaluator = MulticlassClassificationEvaluator(labelCol = 'redshift', metricName = 'accuracy')
print('Decision Tree Accuracy:', multi_evaluator.evaluate(dt_predictions))
print('test data (weightedPrecision): ', multi_evaluator.setMetricName('weightedPrecision').evaluate(dt_predictions))

Decision Tree Accuracy: 0.8618536088466705
test data (weightedPrecision):  0.8672236045024956


In [ ]:
print('test data (weightedPrecision): ', multi_evaluator.setMetricName('weightedPrecision').evaluate(dt_predictions))
#print('test data (weightedRecall): ', multi_evaluator.setMetricName('weightedRecall').evaluate(dt_predictions))

test data (weightedPrecision):  0.8672236045024956


##Cascading

In [ ]:
dt_predictions2 = dt_model.transform(final_df)

In [ ]:
dt_predictions2.show(3)

+--------------------+--------+--------------------+--------------------+----------+
|            features|redshift|       rawPrediction|         probability|prediction|
+--------------------+--------+--------------------+--------------------+----------+
|[21.32289,21.1997...|       2|[26493.0,96884.0,...|[0.15458539744779...|       1.0|
|[21.54044,21.3444...|       0|[1661635.0,16122....|[0.98662245088812...|       0.0|
|[24.09969,21.8938...|       0|[1661635.0,16122....|[0.98662245088812...|       0.0|
+--------------------+--------+--------------------+--------------------+----------+
only showing top 3 rows



In [ ]:
from pyspark.sql.functions import monotonically_increasing_id
dt_predictions = dt_predictions2.withColumn("objid", monotonically_increasing_id())
dt_predictions.show(3)

+--------------------+--------+--------------------+--------------------+----------+-----+
|            features|redshift|       rawPrediction|         probability|prediction|objid|
+--------------------+--------+--------------------+--------------------+----------+-----+
|[21.32289,21.1997...|       2|[26493.0,96884.0,...|[0.15458539744779...|       1.0|    0|
|[21.54044,21.3444...|       0|[1661635.0,16122....|[0.98662245088812...|       0.0|    1|
|[24.09969,21.8938...|       0|[1661635.0,16122....|[0.98662245088812...|       0.0|    2|
+--------------------+--------+--------------------+--------------------+----------+-----+
only showing top 3 rows



In [ ]:
dt_predictions = dt_predictions.drop('redshift')
dt_predictions.show(4)

+--------------------+--------------------+--------------------+----------+-----+
|            features|       rawPrediction|         probability|prediction|objid|
+--------------------+--------------------+--------------------+----------+-----+
|[21.32289,21.1997...|[26493.0,96884.0,...|[0.15458539744779...|       1.0|    0|
|[21.54044,21.3444...|[1661635.0,16122....|[0.98662245088812...|       0.0|    1|
|[24.09969,21.8938...|[1661635.0,16122....|[0.98662245088812...|       0.0|    2|
|[19.70364,17.8579...|[1661635.0,16122....|[0.98662245088812...|       0.0|    3|
+--------------------+--------------------+--------------------+----------+-----+
only showing top 4 rows



In [ ]:
print(df.count(), len(df.columns))
print(dt_predictions.count(), len(dt_predictions.columns))

new_df = df.join(dt_predictions, df.objid == dt_predictions.objid).select(df["*"],dt_predictions["prediction"])
#new_df = df.join(dt_predictions, df.objid == dt_predictions.objid)
new_df.show(5)

3524477 28
3524477 5
+----------+----------+----------+----------+------------------+----------+----------+----------+----------+----------+------------------+----------+------------------+------------------+--------+--------+------------------+--------+-----------+---------+---------+------------------+---------+------------------+---------+---------+--------+-----+----------+
|modelMag_u|modelMag_g|modelMag_r|modelMag_i|        modelMag_z|fiberMag_u|fiberMag_g|fiberMag_r|fiberMag_i|fiberMag_z|        petroR50_r|petroR90_r|        petroR50_z|        petroR90_z|       r|       i|                 z|redshift|       zerr|     mmug|     mmgr|              mmri|     mmiz|              mfug|     mfgr|     mfri|    mfiz|objid|prediction|
+----------+----------+----------+----------+------------------+----------+----------+----------+----------+----------+------------------+----------+------------------+------------------+--------+--------+------------------+--------+-----------+---------+----

In [ ]:
assembler = VectorAssembler(inputCols = ['modelMag_u', 'modelMag_g', 'modelMag_r', 'modelMag_i', 'modelMag_z', 'fiberMag_u', 'fiberMag_g', 'fiberMag_r', 'fiberMag_i', 'fiberMag_z', 'petroR50_r', 'petroR90_r', 'petroR50_z', 'petroR90_z', 'r', 'i', 'z', 'mmug', 'mmgr', 'mmri', 'mmiz', 'mfug', 'mfgr', 'mfri', 'mfiz', 'prediction'], outputCol = 'features2')

In [ ]:
output = assembler.transform(new_df)

In [ ]:
final_df = output.select('features2', 'redshift')
final_df.show(3)

+--------------------+--------+
|           features2|redshift|
+--------------------+--------+
|[21.32289,21.1997...|       2|
|[20.66357,20.0293...|       0|
|[21.01588,20.8195...|       1|
+--------------------+--------+
only showing top 3 rows



In [ ]:
train, test = final_df.randomSplit([0.7, 0.3])

In [ ]:
from pyspark.ml.classification import (DecisionTreeClassifier, RandomForestClassifier, 
                                      GBTClassifier)
from pyspark.ml import Pipeline

dt1 = DecisionTreeClassifier(labelCol = 'redshift', featuresCol = 'features2')
rf = RandomForestClassifier(labelCol = 'redshift', featuresCol = 'features2')
gb = GBTClassifier(labelCol = 'redshift', featuresCol = 'features2')

In [ ]:
start_time = time.time()
dt_model1 = dt1.fit(train)
print("%s seconds" % (time.time() - start_time))

284.2905762195587 seconds


In [ ]:
dt_train_prediction2 = dt_model1.transform(train)
print('DT Training Accuracy:', multi_evaluator.evaluate(dt_train_prediction2))
print('Training data (weightedPrecision): ', multi_evaluator.setMetricName('weightedPrecision').evaluate(dt_train_prediction2))

Random Forest Training Accuracy: 0.8700843556128427
Training data (weightedPrecision):  0.8700843556128427


In [ ]:
dt_predictions2 = dt_model1.transform(test)

In [ ]:
from pyspark.ml.evaluation import MulticlassClassificationEvaluator

multi_evaluator = MulticlassClassificationEvaluator(labelCol = 'redshift', metricName = 'accuracy')
print('Decision Tree Accu:', multi_evaluator.evaluate(dt_predictions2))
print('test data (weightedPrecision): ', multi_evaluator.setMetricName('weightedPrecision').evaluate(dt_predictions2))

Decision Tree Accu: 0.8725439814354761
test data (weightedPrecision):  0.8699307126975936


In [ ]:

#print('test data (weightedRecall): ', multi_evaluator.setMetricName('weightedRecall').evaluate(dt_predictions2))

#Before Cascading
Training Accuracy: 0.8615

Training weightedPrecision: 0.8668

Test Accuracy: 0.8618

Test weightedPrecision: 0.8672

#After Cascading
Training Accuracy: 0.8700

Training weightedPrecision: 0.8700

Test Accuracy: 0.8725

Test weightedPrecision: 0.8699


### Training Time: 